### DISCUSSION TITLE
Help with Digital Ocean Spaces Object Storage

### DISCUSSION THREAD

**Discussion by MarkoTechno** [HiEventsDev/Hi.Events#95] [General]
I'm grateful that this project exists! I have been working to get Hi.Events running in production on Digital Ocean, and having had little experience with droplets, spaces, etc., it's been a learning curve- I've mostly got it working so first I'll share what I've done in case it helps others, and then I have questions.

Here are the environment variables I changed in order to go from local storage to spaces object storage:

```
FILESYSTEM_PUBLIC_DISK: s3-public
FILESYSTEM_PRIVATE_DISK: s3-private
AWS_ACCESS_KEY_ID: {redacted}
AWS_SECRET_ACCESS_KEY: {redacted}
AWS_DEFAULT_REGION: us-east-1
AWS_PUBLIC_BUCKET: my-public-bucket
AWS_PRIVATE_BUCKET: my-private-bucket
AWS_ENDPOINT: https://sfo3.digitaloceanspaces.com
AWS_USE_PATH_STYLE_ENDPOINT: true
```

I created an access key pair on the Digital Ocean API page and specified them in the app environment variables (above). I created 2 spaces buckets (named my-public-bucket and my-private-bucket as examples above) in the sfo3 region where my app is also hosted. I enabled CDN for the public one. Because my app uses a domain name that is not hosted with Digital Ocean, I wanted to use the default URL my-public-bucket.sfo3.cdn.digitaloceanspaces.com - so in order to get the web app to form URLs correctly I had to update the App Spec in Digital Ocean, editing these lines:

```
  - key: APP_CDN_URL
    scope: RUN_TIME
    value: https://my-public-bucket.sfo3.cdn.digitaloceanspaces.com
```

With this change, I could now upload an event image, save it, and the correct URL would be specified in the HTML.

I had to manually change the image permissions via the Digital Oceans Spaces UI, to make the image publicly available. I am the only person who will be creating events on this instance so this is an OK workaround but it'd be nice if that wasn't required.

Two questions:

1. Is there a setting somewhere that will cause the API to set the file permissions to public when the image is uploaded to the Spaces bucket?
2. When does private storage get used? I am not seeing any data being written to the private bucket yet, and I'd like to trigger it to write data, so I can make sure it is working as expected.

Thanks!

**Comment by daveearley** [comment]
Hi @MarkoTechno 👋,

Thanks for trying out Hi.Events.

I hadn't used DigitalOcean Spaces before, but after some debugging I noticed there is indeed a missing setting, which has now been added. You can click "Force rebuild & deploy" to pull down the latest changes.

![Screenshot 2024-07-14 at 10 13 32](https://github.com/user-attachments/assets/d70b2951-11f2-46a6-a294-99eeff4b8336)

I haven't been able to get DigitalOcean Spaces working with their CDN. However, the following settings should work using their Origin Endpoint instead of the CDN endpoint. The CDN returns an SSL error for me, so I believe the issue lies with DigitalOcean and not the app.

```
APP_CDN_URL=https://your-bucket.sfo3.digitaloceanspaces.com
FILESYSTEM_PUBLIC_DISK=s3-public
AWS_ACCESS_KEY_ID=XXX
AWS_SECRET_ACCESS_KEY=XXX
AWS_DEFAULT_REGION=sfo3
AWS_PUBLIC_BUCKET=your-bucket
AWS_URL=https://your-bucket.sfo3.digitaloceanspaces.com
AWS_ENDPOINT=https://sfo3.digitaloceanspaces.com
AWS_USE_PATH_STYLE_ENDPOINT=false
```

Also, the private bucket isn't in use yet, so you can ignore that for now.

Hope this helps.

Cheers,  
Dave

**Reply by MarkoTechno** [reply]
@daveearley thanks for the quick response! I did the force rebuild and deploy, and tweaked the environment variables as suggested. The existing event image continues to work, and more importantly I uploaded a new image and the permissions were set public automatically- thank you!

**Reply by MarkoTechno** [reply]
I'd be happy to take a crack at a PR to document Digital Ocean persistent storage if that helps. Is there a github project for the docs on hi.events? I don't spot them in this project but maybe I'm missing something.

**Reply by daveearley** [reply]
Glad you sorted it! Thanks for the offer, but unfortunately the docs are a work-in-progress and not ready to be made public yet. I'll give you a shout when they are. Cheers!

**Reply by anthonyrom** [reply]
FWIW I just set this up and it seems that both https://your-bucket.sfo3.digitaloceanspaces.com and https://your-bucket.sfo3.cdn.digitaloceanspaces.com resolve to the same thing... so either should work for `APP_CDN_URL`
